java/sage/media/bluray/BluRayFile.java

```diff
@@ -0,0 +1,496 @@
+  {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Unsupported method. This channel can only be read only.
+   *
+   * @throws IOException throws channel is read only.
+   */
+  @Override
+  public long transferFrom(ReadableByteChannel src, long count) throws IOException
+  {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Unsupported method. This channel can only be read only.
+   *
+   * @throws IOException throws channel is read only.
+   */
+  @Override
+  public int write(ByteBuffer src, long position) throws IOException
+  {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Unsupported method.
+   *
+   * @throws IOException throws method is unsupported.
+   */
+  @Override
+  public MappedByteBuffer map(MapMode mode, long position, long size) throws IOException
+  {
+    throw new IOException("Method is unsupported.");
+  }
+
+  /**
+   * Unsupported method.
+   *
+   * @throws IOException throws method is unsupported.
+   */
+  @Override
+  public FileLock lock(long position, long size, boolean shared) throws IOException
+  {
+    throw new IOException("Method is unsupported.");
+  }
+
+  /**
+   * Unsupported method.
+   *
+   * @throws IOException throws method is unsupported.
+   */
+  @Override
+  public FileLock tryLock(long position, long size, boolean shared) throws IOException
+  {
+    throw new IOException("Method is unsupported.");
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException
+  {
+    int readBytes = sageFileChannel.read(dst, position - fileOffsets[currFileIndex]);
+
+    if (dst.hasRemaining())
+    {
+      ensureProperFile(false);
+      if (readBytes == -1)
+        readBytes = 0;
+      readBytes += sageFileChannel.read(dst, position + readBytes - fileOffsets[currFileIndex]);
+    }
+
+    return readBytes;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException
+  {
+    int readBytes = sageFileChannel.read(dst);
+
+    if (dst.hasRemaining())
+    {
+      ensureProperFile(false);
+      if (readBytes == -1)
+        readBytes = 0;
+      readBytes += sageFileChannel.read(dst);
+    }
+
+    return readBytes;
+  }
+
+  @Override
+  public int readUnsignedByte() throws IOException
+  {
+    return sageFileChannel.readUnsignedByte();
+  }
+
+  /**
+   * Unsupported method. This channel can only be read only.
+   *
+   * @throws IOException throws channel is read only.
+   */
+  @Override
+  public int write(ByteBuffer src) throws IOException
+  {
+    throw new NonWritableChannelException();
+  }
+
+  @Override
+  public long position()
+  {
+    return sageFileChannel.position() + fileOffsets[currFileIndex];
+  }
+
+  @Override
+  public long skip(long n) throws IOException
+  {
+    long seek = Math.min(position() + n, totalSize);
+    ensureProperFile(false, seek);
+    return sageFileChannel.skip(n - fileOffsets[currFileIndex]);
```

Comment:
This isn't right because if it doesn't open a new file, it'll skip byte from the file start rather than the offset. To simplify this, I'd say just call position(seek).